Aligned AR and Evidence ASN.1, added text about Endorsements (#201)
lamps-wg/csr-attestation: draft-ietf-lamps-csr-attestation.md

--- middle

# Introduction

When requesting a certificate from a Certification Authority (CA), a PKI end entity may wish to include attestation data (typically Evidence, Endorsements and
Attestation Results) of the security properties of its environments in which the private keys are stored in that request.

Evidence and Endorsements are appraised by Verifiers, which typically produces Attestation Results that serve as input for validating incoming certificate requests against specified certificate policies.
Verifiers are associated with Registration Authorities (RAs) or CAs and function as logical entities responsible for processing Evidence and Endorsements in order to produce Attestation Results.
As remote attestation technology matures, it is natural for a Certification Authority to rely on remote attestation data for proof that the requesting entity is in a state that matches the certificate profile. This is referred to as the RATS Background Check Model, and is illustrated in {{fig-arch-background}}.

> Alternatively, the attester might have a direct connection to a Verifier to which it presents its Evidence and Endorsements, and receives back an Attestation Result signed by the Verifier which it can include directly in the CSR and save the RA / CA from needing a local Verifier. This is referred to as the RATS Passport Model, and is illustrated in {{fig-arch-passport}}.

Reviewer: ```suggestion
Alternatively, the Attester might have a direct connection to a Verifier to which it presents its Evidence and Endorsements, and receives back an Attestation Result signed by the Verifier which it can include directly in the CSR and save the RA / CA from needing a local Verifier. This is referred to as the RATS Passport Model, and is illustrated in {{fig-arch-passport}}.
```

At the time of writing, the most pressing example of the need for remote attestation in certificate enrollment is the Code-Signing Baseline Requirements (CSBR) document maintained by the CA/Browser Forum {{CSBR}}, which requires compliant CAs to "ensure that a Subscriber’s Private Key is generated, stored,
and used in a secure environment that has controls to prevent theft or misuse", which is a natural fit to enforce via remote attestation.

> This specification defines an attribute and an extension that allow for conveyance of Evidence, Endorsements and Attestation Results in Certificate Signing Requests (CSRs), such as PKCS#10 {{RFC2986}} or Certificate Request Message Format (CRMF) {{RFC4211}} payloads.

Reviewer 2: ```suggestion
This specification defines an attribute and an extension that allow for conveyance of Evidence, Endorsements, and Attestation Results in Certificate Signing Requests (CSRs), such as PKCS#10 {{RFC2986}} or Certificate Request Message Format (CRMF) {{RFC4211}} payloads.
```

> This specification defines an attribute and an extension that allow for conveyance of Evidence Endorsements and a separate attribute and extension for the conveyance of Attestation Results in Certificate Signing Requests (CSRs), such as PKCS#10 {{RFC2986}} or Certificate Request Message Format (CRMF) {{RFC4211}} payloads.

Reviewer: ```suggestion
This specification defines an attribute and an extension that allow for conveyance of Evidence, Endorsements and Attestation Results in Certificate Signing Requests (CSRs), such as PKCS#10 {{RFC2986}} or Certificate Request Message Format (CRMF) {{RFC4211}} payloads.
```

This CSR extension satisfies CA/B Forum's CSBR {{CSBR}} requirements for key protection assurance, provided that the CSR carries attestation data that the RA / CA can parse to obtain the assurance that it needs to satisfy its certificate issuance policies.

As outlined in the IETF RATS architecture {{RFC9334}}, an Attester (typically a device) produces a signed collection of Claims that constitute Evidence about its running environment(s).

> The term "attestation" or "remote attestation" is not explicitly defined in RFC 9334 but was later clarified in {{?I-D.ietf-rats-tpm-based-network-device-attest}}.

Reviewer 2: ```suggestion
The terms "attestation" or "remote attestation" are not explicitly defined in RFC 9334 but the activity of "attestation" is clarified in {{?RFC9683}}.
```

It refers to the process of generating and evaluating remote attestation Evidence and Endorsements.

This document relies on {{architecture}} as the foundation for how the various roles within the RATS architecture correspond to a certificate requester and a CA/RA.

Several standard and proprietary remote attestation technologies are in use at the time of writing. This specification thereby is intended to be as technology-agnostic as is feasible with respect to implemented remote attestation technologies. Hence, this specification focuses on (1) the conveyance of Evidence, Endorsements, and Attestation Results via CSRs while making minimal assumptions about content or format of the transported payload and (2) the conveyance of sets of certificates used for validation of Evidence, Endorsements, and Attestation Results.

> The certificates typically contain one or more certification paths rooted in a device manufacturer trust anchor and the end entity certificate being on the device in question. The end entity certificate is associated with key material that takes on the role of an Attestation Key and is used to sign Evidence originating from the Attester. Some remote attestation technologies consider the Attestation Key and its corresponding certificate chain to be Endorsements of the Attestation Key. For the purposes of this document, this is considered to be the certificate chain, and the word "Endorsement" is reserved to refer to a complete set of Claims about the target environment, signed by a 2nd or 3rd party.

Reviewer 2: I assume that Trust Anchors are a minimal Endorsement, based on [Endorsing Verification Keys](https://www.ietf.org/archive/id/draft-ietf-rats-endorsements-07.html#name-endorsing-verification-keys).

> The certificates typically contain one or more certification paths rooted in a device manufacturer trust anchor and the end entity certificate being on the device in question. The end entity certificate is associated with key material that takes on the role of an Attestation Key and is used to sign Evidence originating from the Attester. Some remote attestation technologies consider the Attestation Key and its corresponding certificate chain to be Endorsements of the Attestation Key. For the purposes of this document, this is considered to be the certificate chain, and the word "Endorsement" is reserved to refer to a complete set of Claims about the target environment, signed by a 2nd or 3rd party.

Author: My intended purpose of this new text is to clarify that if you are transmitting a chain of certificates in order to verify the Evidence, then that belongs in `EvidenceBundle.certs` field, not in a separate Endorsement object; even if some standards will refer to these certs an "Endorsement"

The text that I wrote is not clear and should be improved.

> The certificates typically contain one or more certification paths rooted in a device manufacturer trust anchor and the end entity certificate being on the device in question. The end entity certificate is associated with key material that takes on the role of an Attestation Key and is used to sign Evidence originating from the Attester. Some remote attestation technologies consider the Attestation Key and its corresponding certificate chain to be Endorsements of the Attestation Key. For the purposes of this document, this is considered to be the certificate chain, and the word "Endorsement" is reserved to refer to a complete set of Claims about the target environment, signed by a 2nd or 3rd party.

Author: New text:

> In some interpretations of the RATS Architecture {{RFC9334}}, the Attestation Key Certificate and its corresponding certificate chain are considered to be Endorsements of the Attestation Key. In this specification, any data field for carryng a signed object, Evidence, Endorsement, or AttestationResult, is accompanied by a Certs field for carrying any certificate(s) needed to validate the signed object. For the purposes of this specification, a certificate chain provided for the purposes of validating another signed object is not considered to be an Endorsement in and of itself.

- Evidence and Endorsements are placed into an EvidenceStatement along with an OID to identify its type and optionally a hint to the Relying Party about which Verifier (software package, a microservice or some other service) will be capable of parsing it. A set of EvidenceStatement structures may be grouped together along with the set of CertificateChoice structures needed to validate them to form a EvidenceBundle.

- Attestation Results are carried in the AttestationResult along with an OID to identify its type. A set of AttestationResult structures may be grouped together to form an AttestationResultBundle.

A CSR may contain one or more Evidence payloads. For example Evidence
asserting the storage properties of a private key, Evidence
asserting firmware version and other general properties
of the device, Evidence signed using different cryptographic
algorithms, or Endorsements provided by the device manufacturer.

> Like-wise a CSR may also contain one or more Attestation Result payloads.

Reviewer 2: ```suggestion
Like-wise, a CSR may also contain one or more Attestation Result payloads.
```

The certificates typically contain one or more certification paths rooted in a device manufacturer trust anchor and the end entity certificate being on the device in question. The end entity certificate is associated with key material that takes on the role of an Attestation Key and is used to sign Evidence originating from the Attester. In some interpretations of the RATS Architecture {{RFC9334}}, the Attestation Key Certificate and its corresponding certificate chain are considered to be Endorsements of the Attestation Key. In this specification, all data fields intended for carryng a signed object, Evidence, Endorsement, or AttestationResult, is accompanied by a Certs field for carrying any certificate(s) needed to validate the signed object. For the purposes of this specification, a certificate chain provided for the purposes of validating another signed object is not considered to be an Endorsement in and of itself. Here, the term "Endorsement" means a signed object containing data about the target environment which may or may not be accompanied by a certificate chain.

> 

Author: I touched that up a little and deleted the sentence in question, because I agree it was unnecessary and confusing.

- Evidence and Endorsements are placed into an EvidenceStatement along with an OID to identify its type and optionally a hint to the Relying Party about which Verifier (software package, a microservice or some other service) will be capable of parsing it. A set of EvidenceStatement structures may be grouped together along with the set of CertificateChoice structures needed to validate them to form a EvidenceBundle.

- Attestation Results are carried in the AttestationResult along with an OID to identify its type. A set of AttestationResult structures may be grouped together to form an AttestationResultBundle.

A CSR may contain one or more Evidence payloads. For example Evidence
asserting the storage properties of a private key, Evidence
asserting firmware version and other general properties
of the device, Evidence signed using different cryptographic
algorithms, or Endorsements provided by the device manufacturer.

> Like-wise, a CSR may also contain one or more Attestation Result payloads.

Reviewer 3: Question:  Who has to interpret the Attestation Result payload?  Unlike the Evidence (and apparently Endorsement), I believe an AR needs to be something that the CA can interpret as it is a signed result from a Verifier.

You need ONE and only one format for an AttestationResult, not one per Verifier type.    I would recommend either removing AttestationResult and its ilk from the document - OR - actually designing a single format for an AttestationResult.  Given that ARs are creeping features and not part of the original adoption approval, and no one has even taking a swing at a single AR format,  the former is more likely to get through WG last call.

> Like-wise, a CSR may also contain one or more Attestation Result payloads.

Author: Disagree.
I agree that AR formats -- much like Evidence formats -- are currently a Wild Wild West, but I disagree that  that should prevent this document from specifying, generically, how to carry an AR, or that this document should pick a single format.

This document just specifies a generic envelope. Regardless, implementers of attesters will need to do some interop testing with CAs. The goal of this document has never been to fully solve interop for all of attestation. So I disagree with this suggestion.

                            |                 | Policy
                            '------------+----'
                                 ^       |
                      Evidence / |       | Attestation

>                       Endorsmts  |       | Result

Reviewer 2: Maybe re-arrangement of the ascii art is better than Endorsmts?

Most of these aspects are, however, outside the scope of this specification but relevant
for use with a given attestation technology.

The focus of this specification is on the transport of Evidence, Endorsements, and Attesation Results
from the Attester to the Relying Party via existing CSR messages.

# Information Model

## Model for Evidence and Endorsements in CSR

To support a number of different use cases for the transmission of
Evidence, Endorsements and certificate chains needed to verify them in a CSR the structure
shown in {{fig-info-model}} is used.

On a high-level, the structure is composed as follows:
A PKCS#10 attribute or a CRMF extension contains one
EvidenceBundle structure. The EvidenceBundle contains one or more
EvidenceStatement structures as well as one or more
CertificateChoices which enable to carry various format of
certificates. For the purpose of conveyance within these structures,
Evidence and Endorsements are considered interchangeable since they are both signed data objects with a certificate chain that needs to be validated by a Verifier, so for the

> remainder of this document, the term "Evidence" will be used to refer to both types of remote attestation messages.

Reviewer 2: ```suggestion
remainder of this document, the term "Evidence" will be used to refer to both types of RATS conceptual messages.
```

value of stmt.

~~~

> EvidenceBundle ::= SEQUENCE {

Author: oops. Copy&Paste error.
